### ISSUE TITLE
Rcpp::stop() crashes R on linux 

### ISSUE THREAD

**Issue by rplzzz** [JGCRI/hector#314]
To deal with cases where Hector throws an exception in the R interface, we catch the exception, capture the error message, and call `Rcpp::stop()`.  Here's an example:
https://github.com/JGCRI/hector/blob/01c069b60edac0766153e46f9c2c328486c60baa/src/rcpp_hector.cpp#L177-L181

On OS X (or possibly just when built with Clang, this has the expected behavior:
```
> hc <- newcore('inst/input/hector_rcp85.ini')
> run(hc, 5000)
Error in run(hc, 5000) : 
  Error while running hector:  msg:  	Assertion failed: bad Tk value
func: 	ocean_csys_run
file: 	ocean_csys.cpp
ffile:	ocean_csys.cpp

line: 	130
>
```
The R session continues as expected.

```
> hc <- newcore('inst/input/hector_rcp85.ini')
> run(hc, 5000)
*** glibc detected *** /share/apps/R/3.3.3/lib64/R/bin/exec/R: free(): invalid pointer: 0x0000000005f13bb8 ***
======= Backtrace: =========
/lib64/libc.so.6[0x3407c75f4e]
/lib64/libc.so.6[0x3407c78cad]
/people/link593/R/x86_64-pc-linux-gnu-library/3.3/hector/libs/hector.so(_Z31exception_to_condition_templateIN4Rcpp9exceptionEEP7SEXPRECRKT_b+0x1da)[0x7f0a5acb0b0e]
/people/link593/R/x86_64-pc-linux-gnu-library/3.3/hector/libs/hector.so(_Z29rcpp_exception_to_r_conditionRKN4Rcpp9exceptionE+0x29)[0x7f0a5acafa61]
/people/link593/R/x86_64-pc-linux-gnu-library/3.3/hector/libs/hector.so(_hector_run+0x325)[0x7f0a5acac012]
/share/apps/R/3.3.3/lib64/R/lib/libR.so(+0xdc548)[0x7f0a640c1548]
(and a bunch more such stuff)
```
The effect here is to crash the entire R session, which is very bad.

I had hoped that this was just an artifact of a broken configuration on PIC, but when I added a test specifically for catching hector exceptions, it looks like it's happening on Travis as well.
```

> library(testthat)
> library(hector)
> 
> test_check("hector")
 *** caught segfault ***
address (nil), cause 'unknown'
Traceback:
 1: (function (core) {    .Call("_hector_shutdown", PACKAGE = "hector", core)})(<environment>)
An irrecoverable exception occurred. R is aborting now ...
Segmentation fault (core dumped)
```
This is a little weird, in that it appears to have happened while R was trying to clean up the hector core, but I'm not sure if that's significant or not.

The other weird thing is that I'm pretty sure that this wasn't a problem back in the fall when we were running Monte Carlo runs on PIC.  

I'm going to see if I can isolate the problem in a mini-package that doesn't involve hector.  If I can, then I'll post a bug report and see if I can get some answers.  Otherwise, we're going to have to find some kind of workaround.

**Comment by rplzzz** [comment]
I got some responses on Stack Overflow about this, and the consensus seems to be that this is entirely caused by using an “ancient” (i.e., two years old) version of R.  I’m not sure this explains why the test suite started failing on Travis when I added a test for catching Hector exceptions, but one thing at a time.  I’ve constructed a minimal test package: https://github.com/rplzzz/rcpptest .  I’ve been able to verify that this crashes under R 3.3.3 on PIC and succeeds under R 3.4.3 with g++ v 8.1.0.  However, there are a lot of moving parts here, and I suspect that the problem may well have been due to R packages being built with inconsistent versions of g++.  That would mean that everything actually _could_ run on R 3.3.3, if I went back and rebuilt all of the installed packages, making sure to use g++ 8.1 for all of them.

This seems to have fixed the problem where exceptions due to invalid parameters were causing the entire R session to crash, but our tests still weren’t passing.  I’ll spare you the details, but after a great deal of trial and error, it looks as though the very long hector run I was doing to force an error was causing some sort of memory corruption (again, only on linux/g++).  I replaced that test with one where we set an invalid parameter and try to run with it (which is more like the kinds of errors we want to catch), and that seems to have cleared up the problem.

The one thing I _didn’t_ do is to bump the R version requirement to v. 3.4.  For one thing, I’m not entirely convinced that it’s necessary.  For another, I’m still using version 3.3.3 on my system, and I’d like to at least get through this next set of papers before undertaking the arduous upgrade process.

Anyway, long story short, Hector exceptions are now being caught and handled on all three platforms.  It does, however, appear, that on some platforms very long Hector runs (through 5000 AD) create some memory corruption that on some platforms causes a crash when you try to clean up the Hector core.  That seems like a problem for another time, so I've created that as a separate issue #316.

**Comment by bpbond** [comment]
Interesting. Thanks for writing this up @rplzzz .
